kvserver,storage: Update snapshot strategy to use shared storage

itsbilal opened the pull request:
If the sender node was created with a SharedStorage, switch to fast ingestion where we ScanInternal() the keys not in shared levels, and just share the metadata for files in shared levels. The sender of the snapshot specifies in the Header that it is using this ability, and the receiver rejects the snapshot if it cannot accept shared snapshots.

If ScanInternal() returns an `ErrInvalidSkipSharedIteration`, we switch back to old-style snapshots where the entirety of the range is sent over the stream as SnapshotRequests.

Future changes will add better support for detection of when different nodes point to different blob storage buckets / shared storage locations, and incorporate that in rebalancing.

Fixes #103028.

Commit:
kvserver,storage: Update snapshot strategy to use shared storage

If the sender node was created with a SharedStorage, switch to
fast ingestion where we ScanInternal() the keys not in shared
levels, and just share the metadata for files in shared levels.
The sender of the snapshot specifies in the Header that it
is using this ability, and the receiver rejects the snapshot if
it cannot accept shared snapshots.

If ScanInternal() returns an `ErrInvalidSkipSharedIteration`,
we switch back to old-style snapshots where the entirety
of the range is sent over the stream as SnapshotRequests.

Future changes will add better support for detection of when
different nodes point to different blob storage buckets / shared
storage locations, and incorporate that in rebalancing.

Fixes #103028.

Release note (general change): Takes advantage of new CLI option,
`--experimental-shared-storage` to rebalance faster from node to node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105839)
<!-- Reviewable:end -->

itsbilal commented:
Turning into a draft PR until I get to the bottom of the test failures

itsbilal commented:
This is ready for a look now.

bananabrick reviewed (commented):
Took a look at everything but tests, and it looks good, but my mental models for this code are bad. This also deserves a look from someone more familiar with this code, and also someone else from storage.

I'll look at the tests carefully on Monday.

> Reviewed 2 of 40 files at r1, 1 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-:-NZmMcUBAFnixHM7zuTh:bx3s4sf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @pavelkalinnikov)

___
*[`-- commits` line 6 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmOZPM45DpbTCGneX_):*
Unfamiliar with this, are snapshots used for replication? So if we replicate using fast ingestion, with a RF=3, will the actual data only be present on disk once?
___
*[`pkg/cli/debug.go` line 475 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmRR936n8ECud3nnUZ:-NZmRR936n8ECud3nnU_:b-gryqih) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/cli/debug.go#L475)):*
> ```Go
> 
> 	var results int
> 	return rditer.IterateReplicaKeySpans(&desc, snapshot, debugCtx.replicated, rditer.TableKeysInclude, /* skipTableKeys */
> ```

This should be `/* includeTableKeys */`.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 251 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZn6LW1DJYhL5lL0O1j:-NZn6LW1DJYhL5lL0O1k:b-8bfk7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/kv/kvserver/store_snapshot.go#L251)):*
> ```Go
> func (msstw *multiSSTWriter) addRangeDelForLastSpan() error {
> 	if !msstw.doExcise {
> 		panic("this code path is only meant for ")
> ```

Incomplete comment.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 803 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmpMC_ET6JGWsrAaZe:-NZmpMC_ET6JGWsrAaZf:b-5b98d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/kv/kvserver/store_snapshot.go#L803)):*
> ```Go
> 
> 	flushBatch := func() error {
> 		if err := kvSS.sendBatch(ctx, stream, b, ssts, stopFastReplicate, timingTag); err != nil {
> ```

Could we have `stopFastReplicate == true`, but `ssts` also be non-empty?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 893 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmiAbl5g8uh5jQNTpa:-NZmiAbl5g8uh5jQNTpb:b-ca4dgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/kv/kvserver/store_snapshot.go#L893)):*
> ```Go
> 	var valBuf []byte
> 	if fastReplicate {
> 		err := rditer.IterateReplicaKeySpansShared(ctx, snap.State.Desc, snap.EngineSnap, func(key *pebble.InternalKey, value pebble.LazyValue) error {
> ```

The `IterateReplicaKeySpans` function takes a `replicatedOnly` field, which ensures that we are only iterating over fully replicated key spans. But this function doesn't, and it's directly calling `ScanInternal`.  How are we ensuring that we're still only iterating over fully replicated key spans, or do we not need that property?
___
*[`pkg/storage/batch.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZn8Wr7BTTSX5pUU8pO:-NZn8Wr7BTTSX5pUU8pP:b-ok1pcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/storage/batch.go#L204)):*
> ```Go
> 
> // EngineRangeKeys returns the engine range key values at the current entry.
> func (r *BatchReader) EngineRangeKeys() ([]EngineRangeKeyValue, error) {
> ```

super nit: This can use the above function to get the raw keys, and then turn them into engine keys? Feel free to ignore this.


<!-- Sent from Reviewable.io -->

pavelkalinnikov commented on pkg/storage/pebble.go:
```diff
@@ -1434,6 +1440,35 @@ func (p *Pebble) NewEngineIterator(opts IterOptions) EngineIterator {
+
+// ClearRawEncodedRange implements the Engine interface.
+func (p *Pebble) ClearRawEncodedRange(start, end []byte) error {
+	panic("unimplemented")
```
Does this need an implementation? If so, please add a TODO. If not, and it's just to satisfy the interface, could you explain in a comment why this won't be invoked?

pav-kv reviewed (commented):
Flushing some comments.

Is there any way to split this PR into logically independent commits (easier to review)? For example, the part with the "include/exclude user keys" + tests in `rditer` seems nicely separable.

Could you also indicate any particular area Replication reviewers should focus on, or provide tips on how to review this PR (because it's large)?

itsbilal commented:
Thank you so much for the reviews @pavelkalinnikov and @bananabrick. I'll be getting to this shortly; oncall and code-yellow has been occupying more of my time this week.

itsbilal commented:
Thanks for the reviews @pavelkalinnikov and @bananabrick . I just created #107297 which contains two commits: one for the pkg/storage changes, and one for the pkg/kvserver/rditer changes that use those pkg/storage changes. Both of tou should be able to independently review those commits at much greater ease than this mega-PR. I should have also addressed all of your comments that applied to those packages in that PR. I'll now update this PR to address the remainder of those comments, and then rebase it on top of that PR (so it'll be the third commit).

itsbilal commented on pkg/kv/kvserver/kvserverpb/raft.proto:
```diff
@@ -249,16 +249,71 @@ message SnapshotRequest {
+    // If true, the snapshot could contain shared files present in a pre-configured
+    // or explicitly specified shared.Storage instance. Such files will have their
+    // metadata present in the snapshot, but not file contents.
+    bool fast_replicate = 12;
```
Renamed it to `SharedReplicate` - also good with `SharedSnapshot`, let me know if that sounds good. We don't want all uses of shared storage to result in shared replication (eg. non-user keys, and in the future, different S3 buckets eg. multiregion), so something talking about shared _storage_ is probably not ideal.

itsbilal commented on pkg/kv/kvserver/kvserverpb/raft.proto:
```diff
@@ -249,16 +249,71 @@ message SnapshotRequest {
+    InternalKey smallest_point_key = 7;
+    // Largest point key in the sstable. Zero value if no point keys are
+    // present.
+    InternalKey largest_point_key = 8;
```
Each SST has `[smallest,largest]` bounds for point keys, range keys, and overall (which is the union of the point/range key bounds). It mimics the struct in Pebble for `SharedSSTMeta` which in turn relates to `FileMetadata` - the same convention follows through.

itsbilal commented on pkg/kv/kvserver/kvserverpb/raft.proto:
```diff
@@ -249,16 +249,71 @@ message SnapshotRequest {
+  // fast_replicate = true. All contents of this range will be streamed as
+  // usual beyond this point. This bool must be set to true in a request before
+  // the end of the snapshot (i.e. before the final = true request).
+  bool stop_fast_replicate = 6;
```
changed to `stop_shared_replicate`

itsbilal commented on pkg/kv/kvserver/store_snapshot.go:
```diff
@@ -210,6 +218,10 @@ func (msstw *multiSSTWriter) initSST(ctx context.Context) error {
 	}
 	newSST := storage.MakeIngestionSSTWriter(ctx, msstw.st, newSSTFile)
 	msstw.currSST = newSST
+	if msstw.doExcise && msstw.currSpan == len(msstw.keySpans)-1 {
```
Done - moved the comment over to the call to `newMultiSSTWriter` and renamed this variable from `doExcise` to `skipRangeDelForLastSpan` to be more descriptive and refer back to the caller.

itsbilal commented:
Finished updating this PR and addressed all changes suggested so far. Note that the first two commits are #107297 and only the newest (i.e. third) commit should be reviewed here. I'll rebase this branch on master once #107297 merges.

itsbilal commented:
Just rebased on master, this should be a slimmer and easier-to-review PR now.

sumeerbhola reviewed (changes requested):
> Reviewed 2 of 61 files at r5, 1 of 69 files at r7, 3 of 12 files at r9, 13 of 71 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @itsbilal, and @pavelkalinnikov)

___
*[`-- commits` line 11 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW2-h7A7jaT2piXGZV):*
Do we have an existing issue to close this hole of `ErrInvalidSkipSharedIteration`, or at least promote it to the first thing we do in `ScanInternal` before doing the iteration (check the files in the version and their largest seqnum). Assuming we get the error immediately, the caller should have a way to take corrective action and retry. It's not high priority but I'm worried about forgetting about this, and then having an incident where a stray wide local file in L5 caused many snapshots to have to be sent in the old manner.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 342 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWCoSv6LA8EBS7GSQH:-NbWCoSw0ZYAuTk3rCkn:bj9oefd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L342)):*
> ```Go
> 		return errors.AssertionFailedf("start key %s must be before end key %s", end, start)
> 	}
> 	for msstw.keySpans[msstw.currSpan].EndKey.Compare(decodedStart.Key) <= 0 {
> ```

can we lift this sst rollover logic from Put/PutRangeKey/PutInternalPointKey/PutInternalRangeDelete into a separate method?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 363 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWDq7O9YtQ4ZaSenM8:-NbWDq7O9YtQ4ZaSenM9:b-wr6cb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L363)):*
> ```Go
> }
> 
> func (msstw *multiSSTWriter) PutInternalRangeKey(
> ```

PutInternalRangeDelete and PutInternalRangeKey look very similar. Can the code be mostly shared?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 827 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWG9OeAjKc1vCAEyDZ:-NbWG9Of3YOJTvuDpzA3:bxcal03) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L827)):*
> ```Go
> 		b = nil
> 		ssts = ssts[:0]
> 		stopSharedReplicate = false
> ```

why this `stopSharedReplicate=false`?
Is this because it is a single time signal to the receiver to write a rangedel?
It reads oddly since `IterateRKSpansVisitor`, which is the old path, also uses this code. Perhaps rename it as transitionFromSharedToRegularReplicate -- it is understand as a state transition.

Also, I see why this kinda works even though the caller may have written part of the sst before this switch, in that it will now add a RANGEDEL. But RANGEDELs don't apply to data in the same ingested sst, because they have the same seqnum, so maybe it doesn't really work. Is there a unit test for the switchover?
___
*[`pkg/kv/kvserver/kvserverpb/raft.proto` line 260 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW7F_nAhTHFRyU_Vx4:-NbW7F_nAhTHFRyU_Vx5:b-dnq8r4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/kvserverpb/raft.proto#L260)):*
> ```proto
>   }
> 
>   // SharedTable represents one shared SSTable present in shared storage.
> ```

Is this the proto equivalent of `pebble.SharedSSTMeta`? If yes, can you add a comment here that these two should be kept in sync.
___
*[`pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/output` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW8YV-20BzoeEx7g9U:-NbW8YV010DNp3giWtU0:b-o9kq0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c9f44f71afc75f8ee047399f7de81839e3104d1/pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/output#L1)):*
> ```
> echo
> ```

why are rditer testdata files being changed/deleted in this PR?
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 280 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW8kiS7wfuT5uTTFyC:-NbW8kiS7wfuT5uTTFyD:boc6oio) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L280)):*
> ```Go
> 	msstw, err := newMultiSSTWriter(
> 		ctx, cluster.MakeTestingClusterSettings(), scratch, keySpans, 0,
> 		false, /* skipRangeDelForLastSpan */
> ```

is there testing for the true case?


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
TFTR!

> Dismissed @bananabrick and @pavelkalinnikov from 9 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-:-Nbpmq5f8YY7yuYcFZAw:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bananabrick, @pavelkalinnikov, and @sumeerbhola)

___
*[`-- commits` line 6 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmOZPM45DpbTCGneX_):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Unfamiliar with this, are snapshots used for replication? So if we replicate using fast ingestion, with a RF=3, will the actual data only be present on disk once?
</blockquote></details>

Yes, that's right. But S3 is already replicated so we don't need to worry too much there. We do need to worry about S3 availability but that's a whole other can of worms and discussed elsewhere.
___
*[`-- commits` line 11 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW2-h7A7jaT2piXGZV):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Do we have an existing issue to close this hole of `ErrInvalidSkipSharedIteration`, or at least promote it to the first thing we do in `ScanInternal` before doing the iteration (check the files in the version and their largest seqnum). Assuming we get the error immediately, the caller should have a way to take corrective action and retry. It's not high priority but I'm worried about forgetting about this, and then having an incident where a stray wide local file in L5 caused many snapshots to have to be sent in the old manner.
</blockquote></details>

We already iterate over files in L5/L6 as the very first thing in ScanInternal and issue the error if it's going to happen before any keys are surfaced, so that issue is unnecessary. But 1) I didn't want this code to rely too heavily on that for clean abstraction reasons, and 2) We do the ScanInternal at the end of the KV snapshot iteration because the user key range is at the end. Hence the need for `ReplicatedKeysUserOnly` because we could be forced to do a normal iteration on just the user keys
___
*[`pkg/kv/kvserver/store_snapshot.go` line 803 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmpMC_ET6JGWsrAaZe:-NbooOBjCvsp8fLwcy00:b-oqnjen) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/kv/kvserver/store_snapshot.go#L803)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Could we have `stopFastReplicate == true`, but `ssts` also be non-empty?
</blockquote></details>

We throw away sharedSSTs when that happens.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 893 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NZmiAbl5g8uh5jQNTpa:-NbooVi81nN9Yc8WNeX3:b-r8w0ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cd6602fb8bd2e5959c081199175c3c81147133b/pkg/kv/kvserver/store_snapshot.go#L893)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

The `IterateReplicaKeySpans` function takes a `replicatedOnly` field, which ensures that we are only iterating over fully replicated key spans. But this function doesn't, and it's directly calling `ScanInternal`.  How are we ensuring that we're still only iterating over fully replicated key spans, or do we not need that property?
</blockquote></details>

`IterateReplicaKeySpansShared` ensures that. It's not possible for Pebble to know what's replicated and what's not anyway, so that's the best place, abstraction-wise, to ensure it.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 342 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWCoSv6LA8EBS7GSQH:-NbpPzGrDs1z8BYfTwXz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L342)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can we lift this sst rollover logic from Put/PutRangeKey/PutInternalPointKey/PutInternalRangeDelete into a separate method?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 363 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWDq7O9YtQ4ZaSenM8:-NbpPzfvAd5G8P7oTP9S:b-ms51to) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L363)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

PutInternalRangeDelete and PutInternalRangeKey look very similar. Can the code be mostly shared?
</blockquote></details>

Done. (Reduced a lot of code duplication)
___
*[`pkg/kv/kvserver/store_snapshot.go` line 827 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWG9OeAjKc1vCAEyDZ:-NbpU3vD6c7P37a2VC3n:b-1vq1wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L827)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why this `stopSharedReplicate=false`?
Is this because it is a single time signal to the receiver to write a rangedel?
It reads oddly since `IterateRKSpansVisitor`, which is the old path, also uses this code. Perhaps rename it as transitionFromSharedToRegularReplicate -- it is understand as a state transition.

Also, I see why this kinda works even though the caller may have written part of the sst before this switch, in that it will now add a RANGEDEL. But RANGEDELs don't apply to data in the same ingested sst, because they have the same seqnum, so maybe it doesn't really work. Is there a unit test for the switchover?
</blockquote></details>

This works because the caller will get the error before any keys are returned by `IterateReplicaKeySpansShared`, and so we'll throw this flag up and put down the rangedel. The rangedel won't need to apply to keys within the same sst because at that point there won't be any keys in the sst in that span anyway. I've more explicltly called out in the raft.proto comment that this flag must go up before any user keys are streamed. We still need to send a snapshot batch over the wire with this bool as true so the receiver knows to put down the rangedel.

Also did the rename.
___
*[`pkg/kv/kvserver/kvserverpb/raft.proto` line 260 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW7F_nAhTHFRyU_Vx4:-NboqhcS1lzjMCLfcUfV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/kvserverpb/raft.proto#L260)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this the proto equivalent of `pebble.SharedSSTMeta`? If yes, can you add a comment here that these two should be kept in sync.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/output` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW8YV-20BzoeEx7g9U:-Nboq92w89vGa4M2NSBi:b-2ovc9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c9f44f71afc75f8ee047399f7de81839e3104d1/pkg/kv/kvserver/rditer/testdata/TestReplicaDataIterator/r1/all/output#L1)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why are rditer testdata files being changed/deleted in this PR?
</blockquote></details>

It's because `ReplicatedKeysUserOnly` now excludes `ReplicatedByRangeID`, a change made in this PR and not the last one.
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 280 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbW8kiS7wfuT5uTTFyC:-Nbpjetw7jJd9ufNZ99M:byn6k2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L280)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is there testing for the true case?
</blockquote></details>

Done. Now there is.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 61 files at r5, 2 of 12 files at r9, 1 of 71 files at r10, 8 of 11 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-:-NbvyWRB1ig_5ck0wDbE:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bananabrick, @itsbilal, and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 827 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWG9OeAjKc1vCAEyDZ:-NbvxDmY7Euypy5oj7f9:bisvitk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L827)):*
> I've more explicltly called out in the raft.proto comment that this flag must go up before any user keys are streamed.

See comment above about checking this here too.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 265 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbvvUIu3dFXQAvAjdjo:-NbvvUIvDA5lykPZ4El7:b-hgnvp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1075f614073ec92137b073372a3922f134f5cc38/pkg/kv/kvserver/store_snapshot.go#L265)):*
> ```Go
> 	}
> 	// msstw.currSpan == len(msstw.keySpans)-1
> 	if err := msstw.currSST.ClearRawRange(
> ```

Can you add a check that no `Put` call has happened for this span (since that won't be deleted by this rangedel and rangekeydel)?
Or alternatively, since rollover to a new span happens only on a `Put` call, should we be asserting that we are not at the last span?
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 346 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbvyFoc35wHJdlt-ha4:-NbvyFoc35wHJdlt-ha5:b5ssyzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1075f614073ec92137b073372a3922f134f5cc38/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L346)):*
> for the last span which only gets a rangedel when explicitly added.

so this test is only testing the case when it is explicitly added? How about testing the case when `addRangeDelForLastSpan()` is not called?


<!-- Sent from Reviewable.io -->

itsbilal reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-:-Nbz2Q121pn0NMlpPcjp:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bananabrick, @pavelkalinnikov, and @sumeerbhola)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 827 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbWG9OeAjKc1vCAEyDZ:-Nbz2SjE1OBuKpoIGD4C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e44bb30a1af25f0995db174601af9a8017083e31/pkg/kv/kvserver/store_snapshot.go#L827)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> I've more explicltly called out in the raft.proto comment that this flag must go up before any user keys are streamed.

See comment above about checking this here too.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 265 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbvvUIu3dFXQAvAjdjo:-Nbz2R6w-e07UTY1q094:b-x6fz62) ([raw file](https://github.com/cockroachdb/cockroach/blob/1075f614073ec92137b073372a3922f134f5cc38/pkg/kv/kvserver/store_snapshot.go#L265)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add a check that no `Put` call has happened for this span (since that won't be deleted by this rangedel and rangekeydel)?
Or alternatively, since rollover to a new span happens only on a `Put` call, should we be asserting that we are not at the last span?
</blockquote></details>

Done. Went with the latter.
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 346 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/105839#-NbvyFoc35wHJdlt-ha4:-Nbz2TLp6RBbbAcgUc_u:b-eukqsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1075f614073ec92137b073372a3922f134f5cc38/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L346)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> for the last span which only gets a rangedel when explicitly added.

so this test is only testing the case when it is explicitly added? How about testing the case when `addRangeDelForLastSpan()` is not called?
</blockquote></details>

Done. Good catch - added a test case for the other case.


<!-- Sent from Reviewable.io -->

itsbilal commented:
bors r=sumeerbhola
